mautrix/bridge/custom_puppet.py

```diff
@@ -99,9 +124,63 @@ def is_real_user(self) -> bool:
         return bool(self.custom_mxid and self.access_token)
 
     def _fresh_intent(self) -> IntentAPI:
-        return (self.az.intent.user(self.custom_mxid, self.access_token)
+        return (self.az.intent.user(self.custom_mxid, self.access_token, self.base_url)
                 if self.is_real_user else self.default_mxid_intent)
 
+    async def _discover_homeserver_endpoint(self, domain: str) -> str:
+        domain_is_valid = False
+
+        async def validate_versions_api(base_url: str) -> bool:
+
+            async with self.az.http_session.get(urllib.parse.urljoin(base_url, "_matrix/client/versions")) as response:
+                if response.status != 200:
+                    return False
+
+                try:
+                    obj = await response.json(content_type=None)
+                    if len(obj["versions"]) > 1:
+                        return True
+                except (KeyError, json.JSONDecodeError):
+                    return False
+
+        async def get_well_known_homeserver_base_url(probable_domain: str) -> Optional[str]:
+            async with self.az.http_session.get(f"https://{probable_domain}/.well-known/matrix/client") as response:
+                if response.status != 200:
+                    return None
+
+                try:
+                    obj = await response.json(content_type=None)
+                    return obj["m.homeserver"]["base_url"]
+                except (KeyError, json.JSONDecodeError) as e:
+                    raise CouldNotDetermineHomeServerURL(domain) from e
+
+        try:
+            if await validate_versions_api(f"https://{domain}"):
+                # Flag front domain as valid, but keep looking
+                domain_is_valid = True
+        except aiohttp.ClientError:
+            pass
+
+        try:
+            base_url = await get_well_known_homeserver_base_url(domain)
+
+            if base_url is None:
+                if domain_is_valid:
+                    # If we found a valid domain already, we just return that
+                    return f"https://{domain}"
+                else:
+                    raise CouldNotDetermineHomeServerURL(domain)
+
+            if await validate_versions_api(base_url):
+                return base_url
+            elif await validate_versions_api(base_url + "/"):
+                return base_url + "/"
```

Comment:
Related to #20, it's possible this can cause problems

Reply:
Is testing with a slash ever necessary? `validate_versions_api` will already add it if it's missing. It might be necessary to remove the slash if it's there though

Reply:
`validate_versions_api` does `urllib.parse.urljoin(base_url, "_matrix/client/versions")`, which, with an url like `https://homeserver.com/subpath`, can either result in `https://homeserver.com/_matrix/client/versions` being tested, or `https://homeserver.com/subpath/_matrix/client/versions`

Unfortunately, subpath ambiguity is not mentioned in the spec, so if a slash isn't placed, it could very well be both that sub-path, or *not* that subpath